---
title: "Introduction to tidyCoverage"
output: 
  BiocStyle::html_document:
    self_contained: yes
    toc: true
    toc_float: true
    toc_depth: 2
    code_folding: show
date: "`r doc_date()`"
package: "`r pkg_ver('tidyCoverage')`"
vignette: >
  %\VignetteIndexEntry{Introduction to tidyCoverage}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}  
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
    collapse = TRUE,
    comment = "#>",
    crop = NULL, 
    width = 180
)
```

```{r warning = FALSE, include = FALSE, echo = FALSE, message = FALSE, results = FALSE}
library(tidyCoverage)
```

## Introduction 

Genome-wide assays provide powerful methods to profile the composition, the
conformation and the activity of the chromatin. Linear "coverage" tracks 
(generally stored as `.bigwig` files) are one of the outputs obtained 
when processing raw high-throughput sequencing data. These coverage tracks 
can be inspected in genome interactive browsers (e.g. `IGV`) to visually 
appreciate local or global variations in the coverage of specific genomic assays. 

The coverage signal aggregated over multiple genomic features can also be computed. 
This approach is very efficient to summarize and compare the coverage 
of chromatin modalities (e.g. protein binding profiles from ChIP-seq, transcription 
profiles from RNA-seq, chromatin accessibility from ATAC-seq, ...) over 
hundreds and up to thousands of genomic features of interest. This unlocks a more 
quantitative description of the coverage over groups of genomic features. 

`tidyCoverage` implements the `CoverageExperiment` and the `AggregatedCoverage`
classes built on top of the `SummarizedExperiment` class. These classes 
formalize the extraction and aggregation of coverage tracks over 
sets of genomic features of interests. 

## `CoverageExperiment` and `AggregatedCoverage` classes

### `CoverageExperiment`

```{r}
library(tidyCoverage)
showClass("CoverageExperiment")

data(ce)
ce

rowData(ce)

rowRanges(ce)

colData(ce)

assays(ce)

assay(ce, 'coverage')

assay(ce[1, 1], 'coverage')
```

### `AggregatedCoverage`

```{r}
showClass("AggregatedCoverage")

data(ac)
ac

rowData(ac)

rowRanges(ac)

colData(ac)

assays(ac)

assay(ac, 'mean')

assay(ac[1, 1], 'mean')
```

## Manipulate `CoverageExperiment` objects

### Create a `CoverageExperiment` object

```{r}
library(rtracklayer)
bw_file <- system.file("extdata", "MNase.bw", package = "tidyCoverage")
bed_file <- system.file("extdata", "TSSs.bed", package = "tidyCoverage")
CE <- CoverageExperiment(
    tracks = import(bw_file, as = "Rle"), 
    features = import(bed_file),
    width = 3000
)
CE
```

### Bin a `CoverageExperiment` object 

```{r}
coarsen(CE, window = 50)
```

### Expand a `CoverageExperiment` object 

```{r}
expand(CE)
```


### Plot coverage of a set of tracks over a single genomic locus

```{r}
library(ggplot2)

# ~~~~~~~~~~~~~~~ Import coverage tracks into a named list ~~~~~~~~~~~~~~~ #
tracks <- list(
    Scc1 = system.file("extdata", "Scc1.bw", package = "tidyCoverage"), 
    RNA_fwd = system.file("extdata", "RNA.fwd.bw", package = "tidyCoverage"),
    RNA_rev = system.file("extdata", "RNA.rev.bw", package = "tidyCoverage"),
    PolII = system.file("extdata", "PolII.bw", package = "tidyCoverage"), 
    MNase = system.file("extdata", "MNase.bw", package = "tidyCoverage")
) |> BigWigFileList()

locus <- "II:450001-475000"
CE2 <- CoverageExperiment(
    tracks = tracks, 
    features = GRanges(locus),
    width = width   (GRanges(locus))
) |> coarsen(window = width(GRanges(locus))/1000)
CE2 |> 
    expand() |> 
    ggplot(aes(x = coord, y = coverage)) + 
        geom_col(aes(fill = track, col = track)) + 
        facet_grid(track~., scales = 'free') + 
        scale_x_continuous(expand = c(0, 0)) + 
        theme_bw() + 
        theme(legend.position = "none", aspect.ratio = 0.1)
```

## Manipulate `AggregatedCoverage` objects

### Aggregate a `CoverageExperiment` into an `AggregatedCoverage` object

```{r}
AC <- aggregate(CE)
AC

AC10 <- aggregate(CE, bin = 10)
AC10
```

### `AggregatedCoverage` over multiple tracks / feature sets

```{r}
library(purrr)
library(plyranges)

# ~~~~~~~~~~~~~~~ Import genomic features into a named list ~~~~~~~~~~~~~~~ #
features <- list(
    TSSs = system.file("extdata", "TSSs.bed", package = "tidyCoverage"),
    `Convergent transcription` = system.file("extdata", "conv_transcription_loci.bed", package = "tidyCoverage")
) |> map(import) |> map(filter, strand == '+') 

# ~~~~~~~~~~~~~~~ Import coverage tracks into a named list ~~~~~~~~~~~~~~~ #
tracks <- list(
    Scc1 = system.file("extdata", "Scc1.bw", package = "tidyCoverage"), 
    RNA_fwd = system.file("extdata", "RNA.fwd.bw", package = "tidyCoverage"),
    RNA_rev = system.file("extdata", "RNA.rev.bw", package = "tidyCoverage"),
    PolII = system.file("extdata", "PolII.bw", package = "tidyCoverage"), 
    MNase = system.file("extdata", "MNase.bw", package = "tidyCoverage")
) |> map(import, as = 'Rle')

# ~~~~~~~~~~~~~~~ Compute aggregated coverage ~~~~~~~~~~~~~~~ #
CE <- CoverageExperiment(tracks, features, width = 5000, scale = TRUE, center = TRUE)
CE

AC <- aggregate(CE)
AC
```

### Plot aggregated coverages with `ggplot2`

```{r}
as_tibble(AC) |> 
    ggplot(aes(x = coord, y = mean)) + 
    geom_ribbon(aes(ymin = ci_low, ymax = ci_high, fill = track), alpha = 0.2) + 
    geom_line(aes(col = track)) + 
    facet_grid(features ~ .) + 
    labs(x = 'Distance from genomic feature', y = 'Mean coverage') + 
    theme_bw() + 
    theme(legend.position = 'top')
```

### Use a tidy grammar

```{r}
library(tidySummarizedExperiment)
CE |> 
    filter(track != 'Scc1') |> 
    filter(features == 'Convergent transcription') |> 
    aggregate() |> 
    ggplot(aes(x = coord, y = mean)) + 
    geom_ribbon(aes(ymin = ci_low, ymax = ci_high, fill = track), alpha = 0.2) + 
    geom_line(aes(col = track)) + 
    facet_grid(track ~ .) + 
    labs(x = 'Distance from locus of convergent transcription', y = 'RNA-seq coverage') + 
    theme_bw() + 
    theme(legend.position = 'top')
```

### Coarsen the aggregated coverage statistics

```{r}
CE |> 
    filter(track != 'Scc1') |> 
    filter(features == 'Convergent transcription') |> 
    aggregate(bin = 200) |> 
    ggplot(aes(x = coord, y = mean)) + 
    geom_ribbon(aes(ymin = ci_low, ymax = ci_high, fill = track), alpha = 0.2) + 
    geom_line(aes(col = track)) + 
    geom_point(aes(col = track)) + 
    facet_grid(track ~ .) + 
    labs(x = 'Distance from locus of convergent transcription', y = 'Mean coverage') + 
    theme_bw() + 
    theme(legend.position = 'top')
```

## Example use case: `AnnotationHub` and `TxDb` resources

### Recover TSSs of forward human genes 

```{r}
txdb <- TxDb.Hsapiens.UCSC.hg19.knownGene::TxDb.Hsapiens.UCSC.hg19.knownGene
TSSs <- GenomicFeatures::genes(txdb) |> 
    filter(strand == '+') |> 
    anchor_5p() |> 
    mutate(width = 1)
```

### Recover H3K4me3 coverage track from ENCODE

```{r}
library(AnnotationHub)
ah <- AnnotationHub()
ah['AH34904']
H3K4me3_bw <- ah[['AH34904']]
```

### Compute the aggregated coverage of H3K4me3 Â± 3kb around the TSSs of forward human genes 

```{r}
CE <- CoverageExperiment(
    H3K4me3_bw, TSSs, 
    width = 6000, 
    scale = TRUE, center = TRUE
)
AC <- aggregate(CE)
as_tibble(AC) |> 
    ggplot(aes(x = coord, y = mean)) + 
    geom_ribbon(aes(ymin = ci_low, ymax = ci_high, fill = track), alpha = 0.2) + 
    geom_line(aes(col = track)) + 
    facet_grid(track ~ .) + 
    labs(x = 'Distance from TSSs', y = 'Mean coverage') + 
    theme_bw() + 
    theme(legend.position = 'top')
```

### With more genomic tracks

```{r eval = FALSE}
# ~~~~~~~~~~ Recover 15 different histone PTM ChIP-seq tracks ~~~~~~~~~~ #
ids <- c(
    'AH35163', 'AH35165', 'AH35167', 'AH35170', 'AH35173', 'AH35176', 
    'AH35178', 'AH35180', 'AH35182', 'AH35185', 'AH35187', 'AH35189', 
    'AH35191', 'AH35193', 'AH35196'
)
names(ids) <- mcols(ah[ids])$title |> 
    gsub(".*IMR90.", "", x = _) |> 
    gsub("\\..*", "", x = _)
bws <- map(ids, ~ ah[[.x]]) |> 
    map(resource) |> 
    BigWigFileList()
names(bws) <- names(ids)

# ~~~~~~~~~~ Computing coverage over TSSs ~~~~~~~~~~ #
AC <- CoverageExperiment(
    bws, TSSs, 
    width = 4000, 
    scale = TRUE, center = TRUE
) |> aggregate()

# ~~~~~~~~~~ Plot the resulting AggregatedCoverage object ~~~~~~~~~~ #
AC |> 
    as_tibble() |> 
    mutate(
        histone = dplyr::case_when(
            stringr::str_detect(track, 'H2A') ~ "H2A", 
            stringr::str_detect(track, 'H2B') ~ "H2B", 
            stringr::str_detect(track, 'H3') ~ "H3"
        )
    ) |> 
    ggplot(aes(x = coord, y = mean)) + 
    geom_ribbon(aes(ymin = ci_low, ymax = ci_high, fill = track), alpha = 0.2) + 
    geom_line(aes(col = track)) + 
    facet_grid(~histone) + 
    labs(x = 'Distance from TSSs', y = 'Mean histone PTM coverage') + 
    theme_bw() + 
    theme(legend.position = 'top') + 
    hues::scale_colour_iwanthue() + 
    hues::scale_fill_iwanthue() 
```

![](../man/figures/PTMs-TSSs.png)

## Quantify coverage over genomic features

```{r}
# ~~~~~~~~~~ Recover 4 different genomic tracks ~~~~~~~~~~ #
ids <- c(
    'AH35931', # ATAC
    'AH35187', # H3K4me3
    'AH35207', # H3K36me3
    'AH35193'  # H3K9me3
)
names(ids) <- mcols(ah[ids])$title |> 
    gsub(".*IMR90.", "", x = _) |> 
    gsub("\\..*", "", x = _)
bws <- map(ids, ~ ah[[.x]]) |> 
    map(resource) |> 
    BigWigFileList()
names(bws) <- names(ids)

# ~~~~~~~~~~ Define sets of genomic features ~~~~~~~~~~ #
promoters <- GenomicFeatures::promoters(txdb, upstream = 0, downstream = 0) |> unique()
promoters <- promoters[1:2000]

# ~~~~~~~~~~ Quantify coverage over genomic features ~~~~~~~~~~ #
p1 <- CoverageExperiment(bws, promoters, width = 2500, ignore.strand = FALSE, window = 1250) |>
    aggregate() |> 
    ggplot(aes(x = coord, y = mean)) + 
    geom_ribbon(aes(ymin = ci_low, ymax = ci_high, fill = track), alpha = 0.2) + 
    geom_line(aes(col = track)) + 
    labs(x = 'Distance from center of genomic features', y = 'Mean coverage') + 
    facet_grid(~features) + 
    theme_bw() + 
    theme(legend.position = 'top') + 
    geom_vline(xintercept = c(-150, 50), linetype = 'dashed')

# ~~~~~~~~~~ Quantify coverage in NFR region ~~~~~~~~~~ #
NFR <- resize(promoters, width = 200, fix = 'center') |> shift_upstream(50)
upstream <- shift_upstream(NFR, 400)
downstream <- shift_downstream(NFR, 400)
tracks <- list(NFR = NFR, upstream = upstream, downstream = downstream)
p2 <- CoverageExperiment(bws, tracks, width = 200, ignore.strand = FALSE, window = 10) |> 
    S4Vectors::expand() |> 
    group_by(track, features, ranges) |> 
    summarize(coverage = quantile(coverage, probs = 0.9, na.rm = TRUE)) |> 
    ggplot(aes(x = track, y = coverage, fill = features)) + 
    geom_boxplot(outlier.shape = NA) + ylim(c(0, 20)) + 
    labs(x = 'Genomic features', y = 'Mean coverage') + 
    facet_grid(~track, scales = 'free') + 
    theme_bw() + 
    theme(legend.position = 'top')
cowplot::plot_grid(p1, p2)
```

```{r}
TSSs <- system.file("extdata", "TSSs.bed", package = "tidyCoverage") |> import()
tracks <- BigWigFileList(list(
    t05 = "~/Projects/20230517_Lea_MNase-timecourse/results/pneumo_t05_nucpos.bw", 
    t10 = "~/Projects/20230517_Lea_MNase-timecourse/results/pneumo_t10_nucpos.bw", 
    t20 = "~/Projects/20230517_Lea_MNase-timecourse/results/pneumo_t20_nucpos.bw", 
    t40 = "~/Projects/20230517_Lea_MNase-timecourse/results/pneumo_t40_nucpos.bw", 
    t60 = "~/Projects/20230517_Lea_MNase-timecourse/results/pneumo_t60_nucpos.bw"
))

# ~~~~~~~~~~ Quantify coverage over genomic features ~~~~~~~~~~ #
p <- CoverageExperiment(tracks, TSSs, width = 2000, ignore.strand = FALSE) |>
    aggregate() |> 
    ggplot(aes(x = coord, y = mean)) + 
    geom_ribbon(aes(ymin = ci_low, ymax = ci_high, fill = track), alpha = 0.2) + 
    geom_line(aes(col = track)) + 
    labs(x = 'Distance from center of TSSs', y = 'Mean coverage') + 
    theme_bw() + 
    theme(legend.position = 'top') 
```


## Session info

```{r}
sessionInfo()
```
